### ISSUE TITLE
Span authentication support in Jaeger collector

### ISSUE THREAD

**Issue by rabbitstack** [jaegertracing/jaeger#427]
Hi,

As already anticipated [here](https://groups.google.com/forum/#!topic/jaeger-tracing/PcxftflO4_o), we have built a POC around span authentication mechanism. The fork can be found [here](https://github.com/sematext/jaeger). It's nicely integrated with the current Jaeger's code base (the authentication related code is encapsulated inside `security` package, and the authentication can be enabled/disabled via `--collector.auth-span` CLI flag. It currently supports in-memory and SQL-based authentication stores (you can find more details in the google group thread above).

Do you think this is a good starting point and something we/you could iterate on and get into Jaeger? In that case, we could create a PR. Otherwise, if you think this still needs some work,  we would love to see your feedback/suggestion so we can make the appropriated improvements.

Thanks

**Comment by yurishkuro** [comment]
My apologies, I did not have time to review, was traveling the past two week, and need to prep for Velocity NYC next week, but will try to find time this week to review.

**Comment by jpkrohling** [comment]
Thanks for the proposal! Let me start by saying that I really appreciate your efforts and this is a discussion we need to have, sooner or later :)

I'm still a bit uneasy about this, as the whole security and tenancy is, IMO, very connected and would be a better fit for a component deployed side-by-side with Jaeger instead of "as part of" Jaeger. 

From what I can tell, we haven't figured out yet what exactly we want on that front. For instance, *if* we decide that we indeed need security at collector level (and not as a sidecar as I would like to have it), then I'd really vote for having support for JWTs instead of username/password, as it's common for microservices. This way, an external auth server can be configured and the client can obtain an auth token with it. This auth token is then used in the communication with the collector. The collector would only need to know the path to the certificate, to validate the token.

That said, if others are OK with the general approach you've taken, I'd accept it :)

Specifically about your branch, here are a few questions/comments:

- Instead of using "username/password", I would use just use "api-key" or just "token". Using "password" implies, to me, that you are taking care of appropriate storage of its value as if it's a password (salting, peppering and hashing with an appropriate algorithm), which is not the case. Also, it would freak people out to see that the password is in plain text within the span :)

- What's the reasoning behind sending the token within the span? I would rather have the auth to happen on the connection between the tracer and the agent/collector. The existing clients won't be able to authenticate, but I would like to get notified if I have auth problems during my app's bootstrap procedure, instead of wonder why I don't see the spans on the UI :) 

- It would be awesome if the auth store would use the same storage as the span storage. I don't think we can afford to have a DB + Cassandra/ES as a requirement for a production deployment. In fact, I think the `security/authenticationstore` should live at `plugin/storage/{cassandra|es}/authenticationstore`

**Comment by otisg** [comment]
+1 for not using "username/password", which is too specific and sounds like something one should never have in any config file, while API keys/tokens are regularly found in configs.

As for auth store, I think @rabbitstack made it pluggable and in-mem+SQL were just the easiest one to do and others can be added later.

@jpkrohling can you elaborate a bit on what exactly a side-car would be/look like? Would it be another app, another process, running on all machines where apps that are being traced are running?

**Comment by jpkrohling** [comment]
I have nothing against the current auth store implementations (in mem + SQL), I'm just a bit worried that the current state would require both a SQL server and a Cassandra server for production purposes, as we don't have a SQL implementation for the other stores yet. I'd say that it's preferable to use the same storage as the other parts of the collector, but I'll let other expose their opinions.

About the side car, it is indeed a different process running as a proxy in front of the target (collector/agent). On a Kubernetes deployment, this would be a second container on the target's pod, but Kubernetes is not required for this. The approach was discussed a bit here: https://github.com/jaegertracing/jaeger/issues/219#issuecomment-326304370 . On that comment, there are links for @larsmilland's approach and for the blog post on protecting the UI with a Keycloak sidecar.

**Comment by rabbitstack** [comment]
Hi @jpkrohling 

First of all, thank you for your comments.

The username/password approach is left for interoperability (if you scanned the code, the same "auth token" is used for both username as well as password). We wanted to keep this generic for a people that might need that kind of approach, and consequently they would transmit hashed + salted passwords, and we would need to check that too.

I think at this point we could introduce the concept of *authentication strategy*, like JWT token, api-token, username/password, or just keep the "api-key" approach. We agree JWT is also an interesting option.

A running instance of DB shouldn't be a requirement for Jaeger deployment, but rather an option (as you might noticed the SQL query is configurable and can be easily adapted to a specific table structure). Similarly, we could offer LDAP authentication stores, and if the org/users have one they can plugin in and validate the credentials against it. Implementing authentication stores in es/cassandra wouldn't be that hard.

Thanks

**Comment by jpkrohling** [comment]
A salted+hashed password that is transmitted over the wire is just plain text in the end, as that's what will be stored in the final data store. I'm not against the strategy, I'm just in favor of different names, so that people's expectations regarding the storage of this value is met.

About the storage options: I'm sorry for making this more confusing that it really is :) I think you actually agree with me: for production purposes, it shouldn't be expected to use two different storage engines, one for auth and one for "everything else". If the current state of this branch would be sent as PR, it would mean that two storage engines are required for someone using Cassandra as the main backing storage for "everything else". Nothing would prevent them from using SQL for auth and Cassandra for "everything else", though. So, if you'd implement the Cassandra and ES stores on the final PR in addition to the in-mem and SQL, I'd have no concerns.

About LDAP/Auth Strategy/... , this is the rabbit hole I was trying to avoid entering by suggesting the usage of an external auth solution. On the #218, I [exposed](https://github.com/jaegertracing/jaeger/issues/218#issuecomment-311082967) my reasoning, but if most people think that embedding this functionality into the collector is a better approach, I'm OK with that.

**Comment by jpkrohling** [comment]
Something else I forgot to mention: I think it's awesome that you sent this PR! Thank you for that!

**Comment by otisg** [comment]
Thanks for your comments, @jpkrohling !
Re ES/Cassandra - considering Elasticsearch can lose data, would one really want to store auth info there?  I don't know enough about Cassandra to know if it, too, can lose data, and Jepsen Cassandra info is several years old.

Re side-car - I will go read your comments in 218, but my feeling is that side-car increases provisioning/deployment complexity.  That auth server needs to run somewhere.  Somebody needs to provide additional infra and get that deployed.  And there needs to be more than one instance for redundancy.  It's a little easier if one uses containers, but most orgs are not there yet.  I'm also thinking about this whole auth/multi-tenant support in the context of a monitoring SaaS, not just Jaeger deployed within one org.  In such context tenants are really monitoring SaaS users. They typically need to install agents on servers/nodes/VMs they want to monitor, and if they want APM/traces they need to deploy agents attached to their apps.  This is typically easy - a standard install comment - curl ... or rpm... or dpkg... or npm... But if they also need to now install this side-car auth server somewhere, this will be a serious adoption hurdle that will turn all but the most focused and persistent people.  I'll go read your comments now, but I think we also have to keep simplicity of deployment in mind.

**Comment by jpkrohling** [comment]
> I don't know enough about Cassandra to know if it, too, can lose data

I don' think it does, but I'll let the Cassandra experts comment on that. My point is really just about avoiding yet another data storage engine, to be used only by the auth part. Otherwise, we have the same problem as using an auth proxy, as in, it would imply in a more complex provisioning/deployment complexity, requiring HA/redundant storage servers (like we have for Cassandra), ... Which are all really good points, IMO. 

> side-car increases provisioning/deployment complexity

That's true. The simplest solution that works would be to use nginx or Apache's HTTPd as a reverse proxy, taking care of the auth. This is similar (and even simpler) than @LarsMilland's approach. On the other hand, the collector won't have to deal with auth at all, and auth can be as simple or as complex as the deployment requires.

**Comment by otisg** [comment]
> The simplest solution that works would be to use nginx or Apache's HTTPd as a reverse proxy, taking care of the auth.
So tracers would have to send username/password in HTTP headers?
Do current tracers support that?  And do they all send data via HTTP?

**Comment by jpkrohling** [comment]
> So tracers would have to send username/password in HTTP headers?

Yes, some sort of auth (HTTP header, mutual SSL auth, ...) would have be done at the tracer level.

> Do current tracers support that? And do they all send data via HTTP?

Not really: the agent receives data via UDP as well, but I'd assume the agent wouldn't need auth itself. The collector can receive data via thrift or REST, both based on HTTP, so, HTTP auth should work there

**Comment by otisg** [comment]
Aha, OK, so only agent would have to get the auth info into HTTP headers and sent that over to Collector along with traces, so tracers themselves probably don't need to be modified...

Re ES/Cassandra as stores for auth info - there are Apache/Nginx modules for auth against SQL DBs, but nothing like that exists for ES/Cassandra.  Even if we were to write that, most people wouldn't have it in their Apache/Nginx setups.  Because ES can lose data, storing auth data there doesn't feel right (to me).

**Comment by jpkrohling** [comment]
> so tracers themselves probably don't need to be modified

Correct. At least, not in the first moment. We'll eventually need multi tenancy, but we don't know yet how and when, so, this might change in the future.

> Re ES/Cassandra as stores for auth info - there are Apache/Nginx modules for auth against SQL DBs, but nothing like that exists for ES/Cassandra

I'm not quite sure I understand what you mean here. Do you mean that we don't need Cassandra because Apache/nginx doesn't have support for Cassandra as the user storage backend? If our recommended approach is to use Apache/nginx as reverse proxy, the final user won't necessarily require a SQL DB, as the simplest solution that works is to use a htpasswd file as the user storage (not that I'd recommend that either, but that's outside our scope).

>  Because ES can lose data, storing auth data there doesn't feel right (to me).

I'm also not familiar with ES, so, no idea whether we should expect ES to lose data :) But if it does, it certainly doesn't feel right to me as well.

**Comment by otisg** [comment]
> We'll eventually need multi tenancy, but we don't know yet how and when, so, this might change in the future.

Btw. this is what we need and this auth work is just a step towards multi-tenancy support.  That is what we ultimately want and would like to add that very soon in order to avoid forking.

> I'm also not familiar with ES, so, no idea whether we should expect ES to lose data :)

[We know ES quite well](https://sematext.com/consulting/elasticsearch/). It can lose data.  Even its creator has confirmed that.  This is a known thing for ES.

> I'm not quite sure I understand what you mean here. Do you mean that we don't need Cassandra because Apache/nginx doesn't have support for Cassandra as the user storage backend? 

Sort of. I tried pointing out that, yes, Apache/nginx don't have support for doing auth against ES/Cass, but they do have support for doing it against a DB (or flat files).  As we are really aiming for multi-tenancy and auth is just a step in that direction, a DB makes sense as that is where tenant information is most likely to be stored (not in flat files and not in ES/Cass, although that is where the actual data/traces are stored).  So if that sort of multi-tenancy (think performance monitoring SaaS) support is where we are going, I think a DB won't be a problem.

That said, we could adjust this PR and add support for storing auth info ES/Cass if this will help get the PR merged. :)

What say you, @jpkrohling & @yurishkuro ?

**Comment by yurishkuro** [comment]
Just had a brief look at the branch.  I think this ticket could use a clear description of the problem you're trying to solve, e.g. what is the target deployment layout where such mechanism is necessary. If the ultimate goal is multi-tenancy (which, frankly, what I thought the fork was trying to do), then I think we have a bigger discussion ahead of us. The auth-z is just a small piece of it that can be done in different ways, it seems like we're putting the horse in front of the carriage by trying to solve it first.

Side note: the primary change to the core code seems to be adding a filter that can reject spans. Is that correct? We could consider the plugin approach too (rather than forking).

**Comment by jpkrohling** [comment]
>  As we are really aiming for multi-tenancy and auth is just a step in that direction, a DB makes sense as that is where tenant information is most likely to be stored

I would expect the tenant information to be located within a JWT. IMO, Jaeger really doesn't need to know where the user/tenant store is, as Jaeger is potentially only one part of a bigger infra. 

> If the ultimate goal is multi-tenancy (which, frankly, what I thought the fork was trying to do), then I think we have a bigger discussion ahead of us

+1 . We would really benefit from other members of the community sharing their needs and views around multi tenancy. So far, we have one use case (which is really valid!), and we can't decide on an implementation based only on that.

**Comment by rabbitstack** [comment]
>Side note: the primary change to the core code seems to be adding a filter that can reject spans. Is that >correct? We could consider the plugin approach too (rather than forking).
Well, actually is a bit more than that (take a look at the `security` package). It deals with extracting authentication token from the span, delegating the authentication phase to the authentication manager, which in turns lookups for token information inside authentication stores (memory and sql are currently supported), deals with caching to avoid massive round trips to the authentication storage, etc. If you say there's a better approach to accomplish that, we are all ears.

**Comment by otisg** [comment]
> So far, we have one use case (which is really valid!), and we can't decide on an implementation based only on that.

@jpkrohling I have a feeling nobody else is looking at this so I don't think we'll hear about other views.

We can implement the approach where ES or Cassandra can be used as auth stores.  Would that work for both @jpkrohling and @yurishkuro in terms of getting this merged?

> If the ultimate goal is multi-tenancy (which, frankly, what I thought the fork was trying to do), then I think we have a bigger discussion ahead of us. The auth-z is just a small piece of it that can be done in different ways, it seems like we're putting the horse in front of the carriage by trying to solve it first.

Let me start a thread on the mailing list.  I assume that's the best way to start that discussion.

**Comment by ajessup** [comment]
Hey folks, a little late to this thread, but I wonder if [SPIFFE](https://github.com/spiffe/spiffe) could be of service here? SPIFFE is an OSS project that provides PKI for service identity - verifying the identity of, and automatically issuing & rotating PKs and certificates for, running services.

From a client's perspective, the main thing is to be able to use an [SVID](https://github.com/spiffe/spiffe/blob/master/standards/X509-SVID.md) (a specially constructed X.509 certificate), eg. to establish mTLS.

**Comment by jpkrohling** [comment]
@ajessup , would you mind starting a thread on the dev mailing list about this? Right now, I'm playing with abstracting this concern away from Jaeger by using Istio. I'm not familiar with SPIFFE/SVID and would like to keep that in sight.

**Comment by jpkrohling** [comment]
Unfortunately, I couldn't look at SPIFFE in detail yet but looks like it could indeed be a good way to support span authentication. 

The question that remains is whether we want this kind of support inside Jaeger itself, or if auth/z/tenancy can be done at the networking level.
